### PR TITLE
Move relationships from FulfillmentContext to Inherited

### PR DESCRIPTION
**Pull request by spastorino**
r? @lcnr

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/fallback.rs [hunk 1]
```diff
@@ -299,11 +295,13 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
...
-                    relationship.output |= rel.output;
-                }
+            let mut inner_infcx = self.infcx.inner.borrow_mut();
+            let relationships = inner_infcx.find_relationships(root_vid);
+            debug!("relationships: {:#?}", relationships);
+
+            for rel in relationships.iter() {
+                relationship.self_in_trait |= rel.self_in_trait;
+                relationship.output |= rel.output;
             }
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/fallback.rs:305]: Should this loop be moved into `find_relationships`, which would just return a single `FoundRelationships` value?

Comment (spastorino) [reply compiler/rustc_hir_typeck/src/fallback.rs]: Done!

**Review by lcnr — COMMENTED** [review]
looked a bit more into this:

About `relationships`, can you try to instead add the info which is stored there to `TypeVariableValue::Unknown`?

Considering that `FoundRelationships` doesn't really talk about relationships but is info about inference vars that name should make more sense.

**Comment by spastorino** [comment]
**note** this is not ready and the force push didn't work as intended

**Comment by spastorino** [comment]
**note** this is not ready

File: compiler/rustc_hir_typeck/src/inherited.rs [hunk 2]
```diff
@@ -136,6 +140,45 @@ impl<'tcx> Inherited<'tcx> {
         if obligation.has_escaping_bound_vars() {
             span_bug!(obligation.cause.span, "escaping bound vars in predicate {:?}", obligation);
         }
+
+        let infcx = &self.infcx;
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/inherited.rs:144]: I'd prefer if this stays in a separate method.

Also, I think `Inherited` autoderefs to `InferCtxt`, so there's no reason to not just use `self` in place of `infcx.

**Comment by spastorino** [comment]
Just added the last commit but rebased just in case.

**Pull request merged**